### ISSUE TITLE
[LLVM] wrong result from valid bitcode compiled from C source

### ISSUE THREAD

**Issue by zhendongsu** [oracle/graal#4101]
GraalVM version or commit id if built from source: GraalVM CE 22.1.0-dev-20211208_2010 (the latest snapshot build from https://github.com/graalvm/graalvm-ce-dev-builds/releases)
CE or EE: CE
JDK version: openjdk 11.0.13
OS and OS Version: Ubuntu 18.04
Architecture: amd64
The output of `java -Xinternalversion`:
OpenJDK 64-Bit Server VM (11.0.13+8-jvmci-22.0-b02) for linux-amd64 JRE (11.0.13+8-jvmci-22.0-b02), built on Nov 22 2021 22:03:35 by "buildslave" with gcc 7.3.0

```
[520] % clangg22dev -v
GraalVM wrapper script for clang
GraalVM version: Development Build
running: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin/clang -v
clang version 12.0.1 (GraalVM.org llvmorg-12.0.1-3-g6e0a5672bc-bgf11ed69a5a 6e0a5672bc058d882dce3d56f90b72b64a6870d7)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
[521] % 
[521] % lli22dev --version
LLVM 12.0.1 (GraalVM CE Native 22.1.0-dev)
[522] % 
[522] % clangg22dev -c -emit-llvm -o small.bc small.c
[523] % clangg22dev small.c
[524] % ./a.out > ref.txt
[525] % 
[525] % timeout -s 9 5 lli22dev small.bc >& out.txt
Killed
[526] % wc ref.txt out.txt 
       0        1   174776 ref.txt
       0        1 11333632 out.txt
       0        2 11508408 total
[527] % 
[527] % cat small.c
int printf(const char *, ...);
int a, b;
short c(short d, short e) { return d - e; }
int main() {
  for (; a != -10; a = c(a, 3))
    for (b = 0; b < 4; b++)
      printf("%d", 0);
  return 0;
}
```

Expected output: `ref.txt` and `out.txt` should be identical.
However, lli seems to run into an infinite loop and produces incorrect result.

The issue also reproduces with GraalVM 19.3.6 CE.

**Comment by oubidar-Abderrahim** [comment]
Hi, Thank you for reporting this, we will take a look into it and get back to you

**Comment by oubidar-Abderrahim** [comment]
Hi @zhendongsu I tried your example, yet I got a different result
```
wc ref.txt out.txt
     0      1 174776 ref.txt
     0      1 174776 out.txt
     0      2 349552 total
```
out.txt and ref.txt are identical and both execute without infinite loops, this result are obtained using the same GraalVM version you used.

Could you please try de-downloading GraalVM and trying the reproducer example you've provided again?

**Comment by oubidar-Abderrahim** [comment]
I got a confirmation from another team member that he can not reproduce this issue either, this might be related to something different in our setup. Could you please share the compiled `small.bc` file to compare with ours

**Comment by zhendongsu** [comment]
> Could you please share the compiled small.bc file to compare with ours

Yes, attached:
[small.bc.zip](https://github.com/oracle/graal/files/7730263/small.bc.zip)

I notice that the infinite looping is somewhat flaky, so please try to repeat several times.

**Comment by rudihorn** [comment]
I was able to reproduce this bug in the December 15 snapshot, but the underlying issue has since been fixed independently and should the code should run fine in the current master branch.

**Comment by zhendongsu** [comment]
> I was able to reproduce this bug in the December 15 snapshot, but the underlying issue has since been fixed independently and should the code should run fine in the current master branch.

Okay, thanks, Rudi; please feel free to mark it resolved and close the issue. @rudihorn

**Comment by rschatz** [comment]
The fix has also been backported to 22.0, so it should work there, too.
